soweego/importer/imdb_dump_extractor.py

```diff
@@ -0,0 +1,390 @@
+    n_persons = 0
+    n_producers = 0
+    n_writers = 0
+    n_misc = 0
+    n_person_movie_links = 0
+
+    _sqlalchemy_commit_every = 700
+
+    def get_dump_download_urls(self) -> List[str]:
+        """
+        :return: the urls from which to download the data dumps
+        the first URL is the one for the **person dump**, the
+        second downloads the **movie dump**
+        """
+        return [DUMP_URL_PERSON_INFO, DUMP_URL_MOVIE_INFO]
+
+    def _normalize_null(self, entity: Dict) -> None:
+        """
+        IMDB represents a null entry with \\N , this method converts
+        all \\N to None so that they're saved as null in the database.
+        This is done for all "entries" of a given entity.
+
+        The normalization process is done *in place*, so this method
+        has no return value.
+
+        :param entity: represents the entity we want to *normalize*
+        """
+
+        for key, value in entity.items():
+            if value == "\\N":
+                entity[key] = None
+
+    def extract_and_populate(self, dump_file_paths: List[str], resolve: bool) -> None:
+        """
+        Extracts the data in the dumps (person and movie) and processes them.
+        It then proceeds to add the appropiate data to the database. See
+        :ref:`soweego.importer.models.imdb_entity` module to see the SQLAlchemy
+        definition of the entities we use to save IMDB data.
+
+        :param dump_file_paths: the absolute paths of the already downloaded
+        dump files.
+        """
+
+        person_file_path = dump_file_paths[0]
+        movies_file_path = dump_file_paths[1]
+
+        LOGGER.debug("Path to movie info dump: %s", movies_file_path)
+        LOGGER.debug("Path to person info dump: %s", person_file_path)
+
+        start = datetime.datetime.now()
+
+        tables = [
+            imdb_entity.ImdbActorEntity,
+            imdb_entity.ImdbDirectorEntity,
+            imdb_entity.ImdbMovieEntity,
+            imdb_entity.ImdbMusicianEntity,
+            imdb_entity.ImdbProducerEntity,
+            imdb_entity.ImdbWriterEntity,
+            imdb_entity.ImdbPersonMovieRelationship,
+        ]
+
+        db_manager = DBManager()
+        LOGGER.info("Connected to database: %s", db_manager.get_engine().url)
+
+        db_manager.drop(tables)
+        db_manager.create(tables)
+
+        LOGGER.info("SQL tables dropped and re-created: %s",
+                    [table.__tablename__ for table in tables])
+
+        LOGGER.info("Starting to import movies from imdb dump")
+
+        # Here we open the movie dump file, and add everything to the DB
+        with gzip.open(movies_file_path, "rt") as mdump:
+            reader = csv.DictReader(mdump, delimiter="\t")
+
+            # count number of rows for TQDM (so we can display how)
+            # much is missing to complete the process. Then go back
+            # to the start of the file with `.seek(0)`
+            n_rows = sum(1 for line in mdump)
```

Comment:
Nice, but I think this does a linear scan of the whole dump file, thus impacting the time complexity.
Not a big deal, I'm just curious whether you know any more efficient ways to count lines of a text file in Python.

Reply:
As far as I could find only this is the fastest way to count lines in Python. But you're right, it is a bit slow.

```
%%timeit 
dump = gzip.open("20190225_122409.basics.tsv.gz", "rt") 
sum(1 for line in dump)                                                                                                                                                              
> 6.82 s ± 21.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

A faster way might be implementing our own C++ counter, or using `wc` which is much faster:

```
%timeit subprocess.run(["wc", "-l", "20190225_122409.basics.tsv"])
> 237 ms ± 7.71 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
``` 

But this might be a problem if the importer is running on a machine without `wc`. 

If you think 7s is too slow then we can just remove this completely: `tqdm` won't be able to tell us how much is left for the import process to complete, but it will tell us how fast it is going (iterations/second).

Reply:
I was exactly comparing to `wc` in my mind 😄 
Thanks for the detailed answer, my questions was for the sake of curiosity.
7 secs out of an overall lengthy process is just a tiny detail, we can just keep it as is.